File: platform/android-seller/app/src/main/java/ca/shawcliffe/seller/BookingsService.kt
package ca.shawcliffe.seller

import io.github.jan.supabase.auth.auth
import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.engine.okhttp.OkHttp
import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
import io.ktor.client.request.get
import io.ktor.client.request.header
import io.ktor.client.request.parameter
import io.ktor.client.request.patch
import io.ktor.client.request.setBody
import io.ktor.client.statement.HttpResponse
import io.ktor.http.ContentType
import io.ktor.http.contentType
import io.ktor.http.isSuccess
import io.ktor.serialization.kotlinx.json.json
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

// Reads and updates this client's bookings from the web backend, gated
// server-side on the booking_request_system component (same Bearer-token
// pattern as InquiriesService).
object BookingsService {
    @Serializable
    data class BookingCustomer(
        val name: String,
        val phone: String? = null,
        val email: String? = null,
    )

    @Serializable
    data class Booking(
        val id: String,
        val service: String? = null,
        @SerialName("requested_date") val requestedDate: String? = null,
        @SerialName("requested_time") val requestedTime: String? = null,
        val notes: String? = null,
        val status: String,
        @SerialName("created_at") val createdAt: String,
        val customer: BookingCustomer? = null,
    )

    @Serializable
    data class Response(
        val bookings: List<Booking> = emptyList(),
        val total: Int = 0,
        val page: Int = 0,
        val pageSize: Int = 25,
    )

    @Serializable
    private data class StatusUpdate(val status: String)

    private val client = HttpClient(OkHttp) {
        install(ContentNegotiation) { json(Json { ignoreUnknownKeys = true }) }
    }

    suspend fun fetch(status: String? = null, page: Int = 0): Response {
        val token = supabase.auth.currentAccessTokenOrNull() ?: throw BroadcastError.NotAuthenticated

        val response: HttpResponse = client.get("${Config.API_BASE_URL}/api/seller/bookings") {
            header("Authorization", "Bearer $token")
            parameter("page", page)
            if (!status.isNullOrBlank()) parameter("status", status)
        }
        if (!response.status.isSuccess()) {
            throw BroadcastError.Server("Couldn't load bookings.")
        }
        return response.body<Response>()
    }

    suspend fun updateStatus(bookingId: String, status: String) {
        val token = supabase.auth.currentAccessTokenOrNull() ?: throw BroadcastError.NotAuthenticated

        val response: HttpResponse = client.patch("${Config.API_BASE_URL}/api/seller/bookings/$bookingId") {
            header("Authorization", "Bearer $token")
            contentType(ContentType.Application.Json)
            setBody(StatusUpdate(status))
        }
        if (!response.status.isSuccess()) {
            throw BroadcastError.Server("Couldn't update booking.")
        }
    }
}

File: platform/ios-seller/ShawcliffeSeller/BookingsService.swift
import Foundation

// Reads/updates the current client's bookings from the web backend. Mirrors
// InquiriesService's Bearer-token GET pattern against /api/seller/bookings,
// plus a PATCH for status updates.
enum BookingsService {
    struct Response: Decodable {
        let bookings: [Booking]
        let total: Int
        let page: Int
        let pageSize: Int
    }

    private struct StatusUpdate: Encodable {
        let status: String
    }

    private struct ErrorBody: Decodable {
        let error: String?
    }

    static func fetch(status: String? = nil, page: Int = 0) async throws -> Response {
        guard let token = try? await supabase.auth.session.accessToken else {
            throw BroadcastError.notAuthenticated
        }

        var components = URLComponents(
            url: Config.apiBaseURL.appendingPathComponent("api/seller/bookings"),
            resolvingAgainstBaseURL: false
        )!
        var queryItems = [URLQueryItem(name: "page", value: String(page))]
        if let status, !status.isEmpty {
            queryItems.append(URLQueryItem(name: "status", value: status))
        }
        components.queryItems = queryItems

        var request = URLRequest(url: components.url!)
        request.httpMethod = "GET"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            throw BroadcastError.server("Couldn't load bookings.")
        }
        return try JSONDecoder().decode(Response.self, from: data)
    }

    static func updateStatus(bookingId: String, status: String) async throws {
        guard let token = try? await supabase.auth.session.accessToken else {
            throw BroadcastError.notAuthenticated
        }

        var request = URLRequest(
            url: Config.apiBaseURL.appendingPathComponent("api/seller/bookings/\(bookingId)")
        )
        request.httpMethod = "PATCH"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.httpBody = try JSONEncoder().encode(StatusUpdate(status: status))

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw BroadcastError.server("No response from server.")
        }
        guard (200..<300).contains(http.statusCode) else {
            let message = (try? JSONDecoder().decode(ErrorBody.self, from: data))?.error
                ?? "Couldn't update booking (\(http.statusCode))."
            throw BroadcastError.server(message)
        }
    }
}
